api/commands/format.ts

```diff
@@ -0,0 +1,182 @@
+import { constructSyntaxTree, ISyntaxTreeNode } from "@dataform/sqlx/lexer";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as jsBeautify from "js-beautify";
+import * as sqlFormatter from "sql-formatter";
+import { promisify } from "util";
+
+const JS_BEAUTIFY_OPTIONS: JsBeautifyOptions = {
+  indent_size: 2,
+  preserve_newlines: false
+};
+
+export async function formatFile(
+  filename: string,
+  options?: {
+    overwriteFile?: boolean;
+  }
+) {
+  const fileText = await promisify(fs.readFile)(filename, "utf8");
+  const fileExtension = filename.split(".").slice(-1)[0];
+  const formattedText = (() => {
+    try {
+      switch (fileExtension) {
+        case "sqlx":
+          return formatSqlx(constructSyntaxTree(fileText));
+        case "js":
+          return formatJavaScript(fileText);
+        default:
+          return fileText;
+      }
+    } catch (e) {
+      throw new Error(`Unable to format "${filename}": ${e.message}`);
+    }
+  })();
+  if (options && options.overwriteFile) {
+    await promisify(fs.writeFile)(filename, formattedText);
+  }
+  return formattedText;
+}
+
+function formatSqlx(node: ISyntaxTreeNode) {
+  return postProcessFormattedSqlx(
+    getIndividualSqlxStatements(node.contents)
+      .map(individualStatement => {
+        const placeholders: {
+          [placeholderId: string]: ISyntaxTreeNode;
+        } = {};
+        const unformattedPlaceholderSql = individualStatement
+          .map(child => {
+            if (typeof child !== "string") {
+              const placeholderId = crypto.randomBytes(16).toString("hex");
+              placeholders[placeholderId] = child;
+              return placeholderId;
+            }
+            return child;
+          })
+          .join("");
+        return Object.keys(placeholders).reduce(
+          (partiallyFormattedSql, placeholderId) =>
+            formatPlaceholderInSqlx(
+              placeholderId,
+              placeholders[placeholderId],
+              partiallyFormattedSql
+            ),
+          sqlFormatter.format(unformattedPlaceholderSql) as string
+        );
+      })
+      .join("\n\n---\n\n")
+  );
+}
+
+function getIndividualSqlxStatements(nodeContents: Array<string | ISyntaxTreeNode>) {
+  const sqlxStatements: Array<Array<string | ISyntaxTreeNode>> = [[]];
+  nodeContents.forEach(child => {
+    if (typeof child !== "string" && child.startTokenType.endsWith("_statementSeparator")) {
```

Comment:
Can we reference the _statementSeperator from where it's defined

Reply:
fixed as per your other comment